Add HuggingfaceDatasetSplitReader for using Huggingface datasets

pab-vmware opened the pull request:
Added a new reader to allow for reading huggingface datasets as instance
Mapped limited `datasets.features` to `allenlp.data.fields`

Verified for selective dataset and/or dataset configurations for training split, mentioned in the documentation comments of the reader.

Joint work with @divijbajaj @annajung @pkini-vmware @agururajvais

Signed-off-by: Abhishek P (VMware) <pab@vmware.com>

<!-- Please reference the issue number here -->
Fixes #4962

Changes proposed in this pull request:
Introduce a new reader that wraps huggingface datasets to provide instances for a split of the dataset with configuration if required

<!-- To ensure we can review your pull request promptly please ensure ALL GitHub Actions workflows pass -->

pab-vmware commented:
Looking at the _lint_ and _format_ issues. Will address them hopefully today.

pab-vmware commented:
Locally new tests pass - 
`tests/data/dataset_readers/huggingface_datasets_test.py ........
.                                                                                                                                                [ 25%]
`

But since they are doing downloads it is quite possible they fail. Will try to see if I can pin point it. But should be okay for the draft.

dirkgr commented:
It's OK for tests to do downloads, but it's better if the downloads are small. Is there a tiny dataset we can use? Maybe we upload one just for this purpose?

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+        manual_multiprocess_sharding: bool = False,
+        serialization_dir: Optional[str] = None,
+        dataset_name: str = None,
+        split: str = "train",
```
I think it makes the most sense if we use the `file_path` parameter from the reader for `split`. We do this in many other cases too. `file_path` is sort of a legacy name for what that parameter does.

If you do this, you can specify the reader like normal, and then for the training set and the validation set you just put in the name of the split.

dirkgr replied:
It means you have to delay the `load_dataset` call, but that's not a problem, maybe even preferred?

pab-vmware replied:
Makes sense, I can use the `file_path` to read from the specific split indicated.
Let me think this through, I am little rusty on py generator stuff, based on file_path it would need to return from a different split in a generator style. 

Coming to delay of `load_dataset`, we can go in two ways
1. Load all splits during `__init__` it would mean the first `_read` for split will not be slow
2. Load Split on the first `_read` in which case that one read will be really really slow. 

I would prefer 1 over 2, since it is clear but, once Reader object is ready all splits are ready for quick reads. 
Not sure about how _datasets_ handles the memory load, can dig deeper sometime later if required.

dirkgr replied:
If the HF side always downloads all splits, then I can see why you might want to read it in `__init__`. But I suspect it does not. There is no point in downloading a huge training set if you just want to do evaluation. That's why I prefer doing it in `_read()`.

Abhishek-P replied:
We have the control to download split or all splits, just that it returns `Dataset` (one split) or `DatasetDict` (multiple splits) based on the choice. Changing so a split is loaded through hf in reader, introduced a `pre_load ` flag that can be used if required to download all splits in `_init_` if so desired. I can remove that you believe it won't be useful.

pab-vmware replied:
We have the control to download split or all splits, just that it returns `Dataset` (one split) or `DatasetDict` (multiple splits) based on the choice. Changed so a split is loaded through hf in `_read`, introduced a `pre_load ` flag that can be used if required to download all splits in `_init_` if so desired. I can remove that you believe it won't be useful.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+            features = self.dataset.features
+        fields = dict()
+
+        # TODO we need to support all different datasets features described
+        # in https://huggingface.co/docs/datasets/features.html
```
We don't need to support everything right away. If we get 90% of the datasets working, and for 10% we show an intelligible error message, that's already a massive improvement.

pab-vmware replied:
> It's OK for tests to do downloads, but it's better if the downloads are small. Is there a tiny dataset we can use? Maybe we upload one just for this purpose?

Nice Idea, We can actually formulate Dummy dataset to test the specific conversions w.r.t to features -> fields. 
We can add probably sanity scripts to validate specific important datasets to validate the support for said dataset.

pab-vmware replied:
Have added a TODO for this, for now. 
Will take it up in the coming weeks.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                if value.dtype == "string":
+                    # Since TextField has to be made of Tokens add whole text as a token
+                    # TODO Should we use simple heuristics to identify what is token and what is not?
+                    field = TextField([Token(inputs[0][feature])])
```
How about this reader takes a Tokenizer argument, and tokenizes the text accordingly?

pab-vmware replied:
Sure will make that happen.  @divijbajaj had suggested the same. Had kept the scope smaller for First Draft.

pab-vmware replied:
Done.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                    )
+
+            elif isinstance(value, Sequence):
+                # datasets Sequence of strings to ListField of TextField
```
Can you have lists of lists? Should this be recursive?

pab-vmware replied:
Practically yes, did not come across that in these datasets, if I can clean up this mapping formalizations that would help. Make it independent of nesting level.

dirkgr replied:
If the HF dataset spec allows it, I think this should be recursive. Otherwise we'll have a weird problem later when it doesn't work on a dataset that we didn't think of.

pab-vmware replied:
For now, I have added validation for cases that are not supported. 
For supporting nested features I will need to clean up the code a little and it may take me some time.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                    field = ListField(field_list)
+
+            # datasets.Translation cannot be mapped directly
+            # but it's dict structure can be mapped to a ListField of 2 ListField
```
```suggestion
            # but its dict structure can be mapped to a ListField of 2 ListField
```

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                            label=item, label_namespace=feature, skip_indexing=True
+                        )
+                        field_list.append(item_field)
+                    if len(field_list) == 0:
+                        continue
```
Can you have datasets where some instances will have a non-empty list of things, and other instances will have an empty list of things? In that case it seems important to emit the zero-length list anyways, so that batching and padding works correctly with that instance.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+            # datasets.Translation cannot be mapped directly
+            # but it's dict structure can be mapped to a ListField of 2 ListField
+            elif isinstance(value, Translation):
+                if value.dtype == "dict":
```
What happens if it's not a `dict`?

pab-vmware replied:
Documentation for _datasets_ indicates that it will be dict
https://huggingface.co/docs/datasets/package_reference/main_classes.html#datasets.Translation

dirkgr replied:
Can you throw an exception (`ValueError` maybe?) otherwise then, with an explanation? Otherwise we'll see some sort of `NotInitializedException` later, which would be very confusing for a user.

Abhishek-P replied:
Done.

pab-vmware replied:
Done, throwing ValueError indicating `feature` is not supported. Added this as else for similar if/elif elsewhere

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                    texts = list()
+                    for lang in langs:
+                        texts.append(TextField([Token(input_dict[lang])]))
+                    field = ListField([langs_field, ListField(texts)])
```
This works, but would it be difficult to do it in a different way and create two fields, one with the languages, and one with the texts, with a different suffix in their field names?

pab-vmware replied:
Let me try that out.

pab-vmware replied:
Done.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+                    for lang in langs:
+                        index = langs.index(lang)
+                        texts.append(TextField([Token(input_dict["translation"][index])]))
+                    field = ListField([langs_field, ListField(texts)])
```
same question here

pab-vmware replied:
Will try the same

pab-vmware replied:
Done with "-" delimeter.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+            else:
+                raise ValueError(f"Datasets feature type {type(value)} is not supported yet.")
+
+        if field:
```
```suggestion
        if field is not None:
```
There are many things in Python that are falsy. This makes clear which one you mean.

dirkgr commented on setup.py:
```diff
@@ -73,6 +73,7 @@
         "lmdb",
         "more-itertools",
         "wandb>=0.10.0,<0.11.0",
+        "datasets==1.5.0",
```
Do we have to pin the precise version? Can we be at least a little flexible? We don't know what else people's projects will depend on, so giving a little flexibility helps people out.

dirkgr replied:
Cool. Something we do sometimes, like in the `wandb` case above, is pin it to a `major.minor` version, but leave the patch version open. In this case that would be `>=1.5.0,<1.6.0`. Then at least we automatically get little fixes that huggingface provides.

pab-vmware replied:
I have updated this to be the same, hope not testing it with older versions is okay for now.

dirkgr reviewed (changes requested):
I'm really excited to have this. I've been wanting to do this since HF came out with the datasets project.

albertvillanova commented:
Thanks for this PR, @pab-vmware, @divijbajaj, @annajung, @pkini-vmware and @agururajvais.

And thanks @dirkgr for your thorough review.

Just let me know if there is anything we can do on Hugging Face Datasets to help you with the integration. :hugs:

pab-vmware commented:
> It's OK for tests to do downloads, but it's better if the downloads are small. Is there a tiny dataset we can use? Maybe we upload one just for this purpose?

Nice Idea, we can formulate a custom dataset and test around it to validate the mapping of datasets.features to allenlp.data.fields. 

We can probably have sanity scripts to test w.r.t specific important datasets to validate support for them, independent of UT. 

Let me make one such dataset as part of it. Will ask on where to place it (my repo or somewhere else) once I have it.

epwalsh commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+        Reads the dataset and converts the entry to AllenNLP friendly instance
+        """
+        if self.dataset:
+            for entry in self.dataset:
```
To make this efficient with multi-process data loading, consider using `self.shard_iterable()` here like:

```suggestion
            for entry in self.shard_iterable(self.dataset):
```

But I know the `datasets` has a `.select()` method, which you could potentially utilize like:

```suggestion
            for entry in self.dataset.select(self.shard_iterable(range(self.dataset.num_rows))):
```

That might be more efficient, I'm not sure.

pab-vmware replied:
I see `Dataset.select()` doc indicates it makes into a separate dataset. I will go without `select()` for now. Will keep the conv open to re-visit and confirm the better choice before we merge.

pab-vmware replied:
I have used self.shard_iterable. 
Let me add a TODO for checking the select() support

pab-vmware replied:
Have added a TODO for this.

epwalsh commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,190 @@
+    ) -> None:
+        super().__init__(
+            max_instances,
+            manual_distributed_sharding,
+            manual_multiprocess_sharding,
```
If you end up utilizing `self.shard_iterable()` (see my other comment) or similar, you can set these to `True`

```suggestion
            manual_distributed_sharding=True,
            manual_multiprocess_sharding=True,
```

epwalsh reviewed (commented):
This looks really good so far!

epwalsh commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,222 @@
+    """
+
+    def __init__(
+        self,
+        max_instances: Optional[int] = None,
+        manual_distributed_sharding: bool = False,
+        manual_multiprocess_sharding: bool = False,
+        serialization_dir: Optional[str] = None,
+        dataset_name: str = None,
+        config_name: Optional[str] = None,
+        pre_load: Optional[bool] = False
+    ) -> None:
+        super().__init__(
+            max_instances,
+            manual_distributed_sharding=True,
+            manual_multiprocess_sharding=True,
+            serialization_dir,
+        )
```
We usually do this like so:

```suggestion
        self,
        dataset_name: str = None,
        config_name: Optional[str] = None,
        pre_load: Optional[bool] = False,
        **kwargs,
    ) -> None:
        super().__init__(
            manual_distributed_sharding=True,
            manual_multiprocess_sharding=True,
            **kwargs,
        )
```

pab-vmware replied:
Done.

pab-vmware commented:
Looks like I need to do more changes to get it working even for glue. Will inform once it is done.

pab-vmware commented:
Working for 3 datasets. With couple different types of features.

Pending addressing of type check failures

Samples from the 2 datasets, xnli is a little too big for a comment
```
glue-cola:
From:{'idx': 0, 'label': 1, 'sentence': "Our friends won't buy this analysis, let alone the next one we propose."}
To:
Instance with fields:
 	 sentence: TextField of length 1 with text: 
 		[Our friends won't buy this analysis, let alone the next one we propose.]
 
 	 label: LabelField with label: 1 in namespace: 'label'. 
 	 idx: LabelField with label: 0 in namespace: 'idx'. 


_universal_dependencies - en_lines:_
From:
{'deprel': ['mark', 'det', 'nsubj', 'advcl', 'case', 'det', 'compound', 'flat', 'obl', 'case', 'det', 'compound', 'flat', 'nmod', 'punct', 'det', 'nsubj:pass', 'aux:pass', 'root', 'case', 'det', 'compound', 'flat', 'compound', 'obl', 'punct'], 'deps': ['None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None'], 'feats': ["{'Case': 'Nom'}", "{'Definite': 'Ind', 'PronType': 'Art'}", "{'Number': 'Sing'}", "{'Mood': 'Ind', 'Number': 'Sing', 'Person': '3', 'Tense': 'Pres', 'VerbForm': 'Fin'}", 'None', "{'Definite': 'Def', 'PronType': 'Art'}", "{'Number': 'Sing'}", "{'Number': 'Sing'}", "{'Number': 'Sing'}", 'None', "{'Definite': 'Ind', 'PronType': 'Art'}", "{'Number': 'Sing'}", "{'Case': 'Nom'}", "{'Number': 'Sing'}", 'None', "{'Definite': 'Def', 'PronType': 'Art'}", "{'Number': 'Sing'}", "{'Mood': 'Ind', 'Number': 'Sing', 'Person': '3', 'Tense': 'Pres', 'VerbForm': 'Fin'}", "{'Tense': 'Past', 'VerbForm': 'Part', 'Voice': 'Pass'}", 'None', "{'Definite': 'Ind', 'PronType': 'Art'}", "{'Case': 'Nom'}", "{'Case': 'Nom'}", "{'Number': 'Sing'}", "{'Number': 'Sing'}", 'None'], 'head': ['4', '3', '4', '19', '9', '9', '9', '7', '4', '14', '14', '14', '12', '9', '4', '17', '19', '19', '0', '25', '25', '25', '22', '25', '19', '19'], 'idx': 'en_lines-ud-dev-doc1-3177', 'lemmas': ['when', 'a', 'user', 'connect', 'to', 'the', 'SQL', 'server', 'database', 'through', 'a', 'Microsoft', 'Access', 'project', ',', 'the', 'connection', 'be', 'enable', 'through', 'a', 'Windows', 'NT', 'user', 'account', '.'], 'misc': ['None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', "{'SpaceAfter': 'No'}", 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', "{'SpaceAfter': 'No'}", 'None'], 'text': 'When a user connects to the SQL Server database through a Microsoft Access project, the connection is enabled through a Windows NT user account.', 'tokens': ['When', 'a', 'user', 'connects', 'to', 'the', 'SQL', 'Server', 'database', 'through', 'a', 'Microsoft', 'Access', 'project', ',', 'the', 'connection', 'is', 'enabled', 'through', 'a', 'Windows', 'NT', 'user', 'account', '.'], 'upos': [5, 8, 0, 16, 2, 8, 10, 0, 0, 2, 8, 10, 10, 0, 1, 8, 0, 17, 16, 2, 8, 10, 10, 0, 0, 1], 'xpos': [None, 'IND-SG', 'SG-NOM', 'PRES', None, 'DEF', 'SG-NOM', 'SG-NOM', 'SG-NOM', None, 'IND-SG', 'SG-NOM', 'SG-NOM', 'SG-NOM', 'Comma', 'DEF', 'SG-NOM', 'PRES-AUX', 'PASS', None, 'IND-SG', 'SG-NOM', 'SG-NOM', 'SG-NOM', 'SG-NOM', 'Period']}

To:
{'idx': <allennlp.data.fields.text_field.TextField object at 0x147f4b0c0>, 'text': <allennlp.data.fields.text_field.TextField object at 0x147c7a740>, 'tokens': <allennlp.data.fields.list_field.ListField object at 0x145b8c040>, 'lemmas': <allennlp.data.fields.list_field.ListField object at 0x145b8c820>, 'upos': <allennlp.data.fields.list_field.ListField object at 0x145b8c1f0>, 'xpos': <allennlp.data.fields.list_field.ListField object at 0x145b8cd60>, 'feats': <allennlp.data.fields.list_field.ListField object at 0x147ed3a00>, 'head': <allennlp.data.fields.list_field.ListField object at 0x147ed30d0>, 'deprel': <allennlp.data.fields.list_field.ListField object at 0x147ed39d0>, 'deps': <allennlp.data.fields.list_field.ListField object at 0x147ed3a30>, 'misc': <allennlp.data.fields.list_field.ListField object at 0x147ed3af0>}

```

pab-vmware commented:
Have working implementation with limited , features-> field mapping. let me explore getting the code coverage up for the diff. 
I will also look at adding a custom HF dataset so we can use it for testing purposes.

I will try to increase the features that we are mapping but is the current change (post coverage) sufficient to go in?

epwalsh commented:
> I will also look at adding a custom HF dataset so we can use it for testing purposes.

That would great!

dirkgr commented:
What's the status of this? Is it close? Anything you need? I'd love to have this in, maybe even for the 2.3 release!

pab-vmware commented:
Hi Dirk, So now that the coverage is up to 90%, this MR IMO although slightly Raw in w.r.t implementation can go in. 
Let me re-check the known supported dataset/config and update it.

There are things that need to be done, for most of which I have added TODOs and I will pick them up one by one in the coming weeks. 

- Cleanup the code so we have mapper/method for each of the features
- Support Nested Features
- Add Custom Light Dataset and configs and use them fo in UT
- Add a test where the new reader is compared to the customer conll2003 reader - This would validate the reader in a practical manner

pab-vmware commented:
So I have updated the datasets list, added validation tests that are now skipped. Squashed all changes into a single commit. Please take a look when time permits. 

I will work on cleaning up the code and providing nested support. 
If the dataset download during tests becomes problematic, I will prioritize the creation of custom dataset.

dirkgr commented:
> Squashed all changes into a single commit.

Don't worry about doing that for our sake. It'll get squashed when we merge anyways. In fact, now I have no way of seeing just the changes you made since the last time I reviewed this.

dirkgr commented on CHANGELOG.md:
```diff
@@ -8,14 +8,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+=======
+>>>>>>> Fix Doc mistake and the dataset availability check
```
Something happened here during merging?

pab-vmware replied:
Let me handle that.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+    Following dataset and configurations have been verified and work with this reader
+
+            Dataset                       Dataset Configuration
+            `xnli`                        `ar`
+            `xnli`                        `en`
+            `xnli`                        `de`
+            `xnli`                        `all_languages`
+            `glue`                        `cola`
+            `glue`                        `mrpc`
+            `glue`                        `sst2`
+            `glue`                        `qqp`
+            `glue`                        `mnli`
+            `glue`                        `mnli_matched`
+            `universal_dependencies`      `en_lines`
+            `universal_dependencies`      `ko_kaist`
+            `universal_dependencies`      `af_afribooms`
+            `swahili`                     `NA`
+            `conll2003`                   `NA`
+            `dbpedia_14`                  `NA`
+            `trec`                        `NA`
+            `emotion`                     `NA`
+             Note: universal_dependencies will require you to install `conllu` package separately
```
Do you think we'll need this list? I'd sort of hope that when it's done, most datasets will work, especially new ones that don't exist yet today. Then maybe we'd have a list for the exceptions, if at all?

pab-vmware replied:
Right now that is not true, hence I have kept the list. 
When we have larger coverage(most features and nesting support) we would go to a known unsupported datasets.

pab-vmware replied:
We would need to be able to frequently (i.e before each release) test against all HF datasets to do that. 
I can add a script to do that in a follow up .

dirkgr replied:
I don't think we have to test all datasets all the time? Just a few that have some properties that we know make a difference in the code.

pab-vmware replied:
We would need to go through datasets and pick them for those qualities and from what I have seen it will be quite some amount of work. 877 unique ones right now. 

Yes, not adding this list would mean they try it out and raise issue that it does not work. That is good, since it will allow us to improve it. 
Removing this.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+        Name of the dataset from huggingface datasets the reader will be used for.
+    config_name : `str`, optional (default=`None`)
+        Configuration(mandatory for some datasets) of the dataset.
+    preload : `bool`, optional (default=`False`)
+        If `True` all splits for the dataset is loaded(includes download etc) as part of the initialization,
+        otherwise each split is loaded on when `read()` is used for the same for the first time.
```
Why is it important that this is an option? Can't we just always read it right when we need it?

pab-vmware replied:
IMO it is about predictability. Pre-load gives you that it will take time to load in the beginning and when you read it is just that simple read. 

Also this helps making the download happen for multiple splits together if that is plan. 

I think this is a good thing to have. If you think it makes usage complicated, I will remove it.

dirkgr replied:
I think it makes this reader different from all other readers. It would be surprising for users.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+        if preload:
+            self.load_dataset()
+
+    def load_dataset(self):
+        if self.config_name is not None:
+            self.dataset = load_dataset(self.dataset_name, self.config_name)
+        else:
+            self.dataset = load_dataset(self.dataset_name)
```
I don't think we need this in a function of its own. We can just say `dataset = load_dataset(...)` in the `_read()` method. The second time we call this (for the validation or test sets), it'll be fast, because it's cached.

pab-vmware replied:
Will do that.

pab-vmware replied:
Went awat since pre-load was removed.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+        if file_path not in self.dataset:
+            self.load_dataset_split(file_path)
+
+        # TODO see if use of Dataset.select() is better
```
I think it is, because `shard_iterable()` will read everything from the dataset, and then discard what it doesn't need.

Though HF datasets are indexable, right? So you could do this:
```Python
split = self.dataset[file_path]
for index in self.shard_iterable(range(len(split)):
    yield self.text_to_instance(file_path, split[index])
```

That way you don't have to re-implement what `shard_iterable` does.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+    def text_to_instance(self, *inputs) -> Instance:
+        """
+        Takes care of converting dataset entry into AllenNLP friendly instance
+        Currently it is implemented in an unseemly catch-up model
```
I don't think that's unseemly?

pab-vmware replied:
Let me remove the adjective. 
I did think it was unseemly compared to what it should be (recursive with feature specific handlers) hence had added. 
I realise it is not a good thing to have in documentation comment 😅.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+        # TODO we need to support all different datasets features described
+        # in https://huggingface.co/docs/datasets/features.html
+        for feature in features:
+            fields_to_be_added: typing.Dict[str, Field] = dict()
```
You can just say `from typing import Dict` at the top, so you don't have to say `typing.Dict` everywhere.

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+                    self.raise_feature_not_supported_value_error(value)
+
+            # datasets.Translation cannot be mapped directly
+            # but it's dict structure can be mapped to a ListField of 2 ListField
```
This is not accurate anymore, right? Because it now creates two fields with different names?

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+    def raise_feature_not_supported_value_error(self, value):
+        raise ValueError(f"Datasets feature type {type(value)} is not supported yet.")
+
+    def text_to_instance(self, *inputs) -> Instance:
```
Why is this using `*inputs` and not normal parameters?

pab-vmware replied:
Changing that would mean it mismatches with supertype 
`: error: Signature of "text_to_instance" incompatible with supertype "DatasetReader`

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,276 @@
+        features = self.dataset[split].features
+        fields = dict()
+
+        # TODO we need to support all different datasets features described
+        # in https://huggingface.co/docs/datasets/features.html
```
To support arbitrarily nested fields we'll need to have some structure where one function returns an AllenNLP `*Field` (or multiple fields possibly) for a given HF feature. This function will be recursive to support arbitrary nesting. We'll have another function that takes all those created fields and combines them into an AllenNLP `Instance` object.

But how often does that really happen? If we can make this work with 90% of the datasets in HF (and all of the important ones) without properly supporting recursion, I'm OK with it.

pab-vmware replied:
Yes you are right, I think most datasets will use the already implemented features, will remove this TODO. 

When making the list mentioned above, there were quite a few that I had to remove since they use nesting.
I know without nesting support (calling it nesting rather than recursion) 90% will not be supported. 

If you have seen my above comment that is a follow up bullet point.

dirkgr replied:
🤷🏼‍♂️ That sounds like nesting is important then.

pab-vmware replied:
Added Nesting Support for Sequence. 
The only FeatureType expected to have Nesting. 
It is slightly raw since for all datatypes other than string, I am using LabelField

dirkgr reviewed (changes requested):
There are a few typos around, nothing serious.

dirkgr commented:
What's the status of this? It looked like it was pretty close to done?

pab-vmware commented:
I was not able to make time for this for it last couple weeks. Plan to resume this weekend with focus on recursive - mapper based implementation.

pab-vmware commented:
> I think the docstring of line 135 of trainer.py (the model parameter) also needs to be updated
> 
> ```
>     If you are training your model using GPUs, your model should already be
>     on the correct device. (If you are using our `train` command this will be
>     handled for you.)
> ```
> 
> The part in parenthesis implies that the model doesn't have to be on the GPU if you're using the `train` command. I think it should be the same as your change above, i.e.
> 
> ```
>     If you are training your model using GPUs, your model should already be
>     on the correct device, which you can do with `model = model.cuda()`. 
> ```
> 
> Perhaps add a link instead of repeating? i.e.
> 
> ```
>     See `cuda_device` for more details.
> ```
Hi David, 
Not sure which of my change you are referring can you please point me to the File and LineNo.

pab-vmware commented:
Have to re-add some of the inline comments, I have removed during the refactoring. 
I have added mapping Functions as independent functions. 
Let me know if you want me to move them as methods to the reader class.

pab-vmware commented:
> I don't know what's happening here, but GitHub thinks this PR has 8000 lines? It's not reviewable like this. I'll try to find out how to work around that.

Seeing that now. 
Let me try and fix that somehow.

dirkgr commented:
On the one hand, I don't think it's your fault. I re-based the same way, and I also got a mess in the GitHub UI.

On the other hand, just don't re-base (in the future). Just merge `origin/main` into here. It's easier, and it doesn't confuse GitHub.

dirkgr commented:
How did you do it?

dirkgr commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+    item_field: Union[LabelField, TextField]
+    field_list: List[Union[TextField, LabelField]] = list()
+    fields: Dict[str, Field] = dict()
+    if isinstance(value.feature, Value):
+        for item in entry[feature]:
+            # If tokenizer is provided we will use it to split it to tokens
+            # Else put whole text as a single token
+            item_field = map_Value(feature, item, value.feature, tokenizer)
+            field_list.append(item_field)
+        if len(field_list) > 0:
+            fields[feature] = ListField(field_list)
+
+    # datasets Sequence of strings to ListField of LabelField
+    elif isinstance(value.feature, ClassLabel):
+        for item in entry[feature]:
+            item_field = map_to_Label(feature, item, skip_indexing=True)
+            field_list.append(item_field)
+
+        if len(field_list) > 0:
+            fields[feature] = ListField(field_list)
+
+    else:
+        HuggingfaceDatasetReader.raise_feature_not_supported_value_error(value)
```
I haven't looked at Huggingface datasets in enough detail, but I would have expected that there is some recursive call in here. What happens if one sequence feature contains another sequence feature?

pab-vmware replied:
Looks like my claim appears to be wrong as of now. I had a dataset that would fail due to nesting and it passes giving me a false positive. 
I need to add recursion inside `_map_Sequence` with either `_map_Feature` or itself to handle that nesting. 
I will try to formulate that. 

Also I will try to find a dataset which has that so I can make sure it works.

pab-vmware replied:
I have added the required recursion, code looks alright but haven't been able to validate. 
Take a look at the code and let me know if anything wrong pops
I will try to random sample the HF datasets in hub and see if any use sequence of sequence

Looks like sequence of dict is a popular case from a few datasets that start with a. 
I will try to Implement that now.

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+    def raise_feature_not_supported_value_error(value):
+        raise ValueError(f"Datasets feature type {type(value)} is not supported yet.")
+
+    def text_to_instance(self, *inputs) -> Instance:
```
Changing that would mean it mismatches with supertype
: error: Signature of "text_to_instance" incompatible with supertype "DatasetReade

Inheritance.  I believe. Due the nature of the parameter resolution in python there is no _overloading_ of a method.
this means if the child class has a method with the same name, it has to have the same signature.

dirkgr replied:
That's true. This part of the `DatasetReader` API isn't super clean. We do this in other readers, where we just give `text_to_instance()` the parameters it should have, and silence the warning if we get one. I would encourage you do to the same. I hope that we don't have this problem anymore in the next version of the AllenNLP API.

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+        # It would be cleaner to create a separate reader object for diferent dataset
+        if dataset_name not in list_datasets():
+            raise ValueError(f"Dataset {dataset_name} not available in huggingface datasets")
+        self.dataset: DatasetDict = DatasetDict()
```
From a programming perspective having this dict is cleaner in the sense that even datasets gives you a dataset dict. 
And this is not a cache, since the reference is still to the same dataset object given by the datasets lib. 
When a split is loaded it is dataset, for the reader to maintain organization of splits, I am keeping this DatasetDict(

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+            else:
+                self.dataset[split] = load_dataset(self.dataset_name, split=split)
+        else:
+            raise ValueError(
+                f"Only default splits:{self.SUPPORTED_SPLITS} are currently supported."
+            )
```
Basically anyone can have any split with any name. 
But right now I don't know a way to detect all the splits of dataset without actually fully loading it. Hence this. 
I can remove this, but that would mean it would actually fail with the underlying huggingface error.

dirkgr replied:
The underlying huggingface error is something like "Can't find split", or even `KeyError`? I think that's fine.

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+        Reads the dataset and converts the entry to AllenNLP friendly instance
+        """
+        if file_path is None:
+            raise ValueError("parameter split cannot be None")
```
When a user fails to pass it, or passes None explicitly
TBH I experience with shared code libs (limited to 4 years) tells me that, we should handle such mistakes. 
So this check is fail-fast, allowing me to not handle that problem down the line. 
If split is none and I try load it will load the full dataset and things will break.

dirkgr replied:
:+1:

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+        # and another indicate the sentiment (of typeint32/ClassLabel)
+
+        split = inputs[0]
+        features: Dict[str, FeatureType] = self.dataset[split].features
```
Not sure if it is doing that, but I will make it an instance field that is saved during instantion and re-used in this.

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+
+
+# Feature Mappers - These functions map a FeatureType into Fields
+def map_Feature(
```
Done.

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+
+# Feature Mappers - These functions map a FeatureType into Fields
+def map_Feature(
+    feature: str, entry: Dict, value, tokenizer: Optional[Tokenizer]
```
Renamed `feature` to `feature_name`
`entry` is the entry of dataset
`value` is now `feature_type`

pab-vmware commented on allennlp/data/dataset_readers/huggingface_datasets_reader.py:
```diff
@@ -0,0 +1,270 @@
+        fields_to_be_added[feature] = map_Value(feature, entry[feature], value, tokenizer)
+
+    elif isinstance(value, Sequence):
+        fields_to_be_added = map_Sequence(feature, entry, value, tokenizer)
```
Changed now.

pab-vmware commented:
Pending local workflow checks before requesting workflow, here.

pab-vmware commented:
Looks like the refactor did cause regression. the tests I had skipped since they downloaded datasets are not failing for some datasets. Working on fixing them, will keep them unskipped until we are actually ready to merge.

pab-vmware commented:
> Looks like the refactor did cause regression. the tests I had skipped since they downloaded datasets are not failing for some datasets. Working on fixing them, will keep them unskipped until we are actually ready to merge.

Fixed now.

pab-vmware commented:
@dirkgr I am thinking of picking a few datasets and adding clear structure-based test to make sure this implementation is valid to the desired extent. 

1st one I am picking is `conll2003`, Please recommend other datasets so that I can have a clearer goal I can work towards. 
Testing coverage w.r.t to 90% of datasets is not something that is not very easy since I would need to figure out the config for most of the datasets (800+ of them) in addition to having to download them. 

if that is compulsorily required, I will try to write a script to test this for each of the datasets, probably run it in AWS VM I have so as to avoid the network problems. 

For conll2003 - https://huggingface.co/datasets/conll2003
On HF side it is as follows
```
{
    "chunk_tags": [11, 12, 12, 21, 13, 11, 11, 21, 13, 11, 12, 13, 11, 21, 22, 11, 12, 17, 11, 21, 17, 11, 12, 12, 21, 22, 22, 13, 11, 0],
    "id": "0",
    "ner_tags": [0, 3, 4, 0, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    "pos_tags": [12, 22, 22, 38, 15, 22, 28, 38, 15, 16, 21, 35, 24, 35, 37, 16, 21, 15, 24, 41, 15, 16, 21, 21, 20, 37, 40, 35, 21, 7],
    "tokens": ["The", "European", "Commission", "said", "on", "Thursday", "it", "disagreed", "with", "German", "advice", "to", "consumers", "to", "shun", "British", "lamb", "until", "scientists", "determine", "whether", "mad", "cow", "disease", "can", "be", "transmitted", "to", "sheep", "."]
}
```

Currently it is read in the converting to as follows
```

         id: TextField of length 1 with text: 
 		[0]
 
 	 tokens: ListField of 12 TextFields : 
 	 TextField of length 1 with text: 
 		[SOCCER]
 
 	 TextField of length 1 with text: 
 		[-]
 
 	 TextField of length 1 with text: 
 		[JAPAN]
 
 	 TextField of length 1 with text: 
 		[GET]
 
 	 TextField of length 1 with text: 
 		[LUCKY]
 
 	 TextField of length 1 with text: 
 		[WIN]
 
 	 TextField of length 1 with text: 
 		[,]
 
 	 TextField of length 1 with text: 
 		[CHINA]
 
 	 TextField of length 1 with text: 
 		[IN]
 
 	 TextField of length 1 with text: 
 		[SURPRISE]
 
 	 TextField of length 1 with text: 
 		[DEFEAT]
 
 	 TextField of length 1 with text: 
 		[.]
 
 
 	 pos_tags: ListField of 12 LabelFields : 
 	 LabelField with label: 21 in namespace: 'pos_tags'. 
 	 LabelField with label: 8 in namespace: 'pos_tags'. 
 	 LabelField with label: 22 in namespace: 'pos_tags'. 
 	 LabelField with label: 37 in namespace: 'pos_tags'. 
 	 LabelField with label: 22 in namespace: 'pos_tags'. 
 	 LabelField with label: 22 in namespace: 'pos_tags'. 
 	 LabelField with label: 6 in namespace: 'pos_tags'. 
 	 LabelField with label: 22 in namespace: 'pos_tags'. 
 	 LabelField with label: 15 in namespace: 'pos_tags'. 
 	 LabelField with label: 12 in namespace: 'pos_tags'. 
 	 LabelField with label: 21 in namespace: 'pos_tags'. 
 	 LabelField with label: 7 in namespace: 'pos_tags'. 
 
 	 chunk_tags: ListField of 12 LabelFields : 
 	 LabelField with label: 11 in namespace: 'chunk_tags'. 
 	 LabelField with label: 0 in namespace: 'chunk_tags'. 
 	 LabelField with label: 11 in namespace: 'chunk_tags'. 
 	 LabelField with label: 21 in namespace: 'chunk_tags'. 
 	 LabelField with label: 11 in namespace: 'chunk_tags'. 
 	 LabelField with label: 12 in namespace: 'chunk_tags'. 
 	 LabelField with label: 0 in namespace: 'chunk_tags'. 
 	 LabelField with label: 11 in namespace: 'chunk_tags'. 
 	 LabelField with label: 13 in namespace: 'chunk_tags'. 
 	 LabelField with label: 11 in namespace: 'chunk_tags'. 
 	 LabelField with label: 12 in namespace: 'chunk_tags'. 
 	 LabelField with label: 0 in namespace: 'chunk_tags'. 
 
 	 ner_tags: ListField of 12 LabelFields : 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 5 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 1 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
 	 LabelField with label: 0 in namespace: 'ner_tags'. 
```

I see that tokens should be a single TextField of tokens instead of a ListField of TextField. 
Let me focus on getting that working.

dirkgr commented:
Sorry for the delay. I was on vacation for a bit.

For datasets, I would recommend the ones we have models for:
 * SQuAD 1.1 and 2.0
 * CoNLL 2003
 * [sst](https://huggingface.co/datasets/sst)
 * [openwebtext](https://huggingface.co/datasets/openwebtext)
 * MOCHA
 * https://huggingface.co/datasets/commonsense_qa
 * https://huggingface.co/datasets/piqa
 * https://huggingface.co/datasets/swag
 * https://huggingface.co/datasets/snli
 * https://huggingface.co/datasets/multi_nli
 * https://huggingface.co/datasets/glue
 * https://huggingface.co/datasets/super_glue
 * https://huggingface.co/datasets/drop
 
I'm excluding the vision models here because that'll be a different can of worms.

Does that list work for you?

pab-vmware commented:
> Sorry for the delay. I was on vacation for a bit.
> 
> For datasets, I would recommend the ones we have models for:
> 
>     * SQuAD 1.1 and 2.0
> 
>     * CoNLL 2003
> 
>     * [sst](https://huggingface.co/datasets/sst)
> 
>     * [openwebtext](https://huggingface.co/datasets/openwebtext)
> 
>     * MOCHA
> 
>     * https://huggingface.co/datasets/commonsense_qa
> 
>     * https://huggingface.co/datasets/piqa
> 
>     * https://huggingface.co/datasets/swag
> 
>     * https://huggingface.co/datasets/snli
> 
>     * https://huggingface.co/datasets/multi_nli
> 
>     * https://huggingface.co/datasets/glue
> 
>     * https://huggingface.co/datasets/super_glue
> 
>     * https://huggingface.co/datasets/drop
> 
> 
> I'm excluding the vision models here because that'll be a different can of worms.
> 
> Does that list work for you?
Yes, Thanks. Let me start working down this checklist.

pab-vmware commented:
@dirkgr Need a quick advice. When I get a non string simple value, I was mapping it to as a LabelField
But, `skip_indexing` False and True will fail since it is neither `str` nor `int`

I can see the best possible choice other than that would be a Tensor Field. 

e.g. this is coming in the case of

dirkgr commented:
Yes, I think a tensor field is the right thing to do. It can contain a tensor with no dimension:
```
In [5]: t = torch.tensor(2.5)

In [6]: t
Out[6]: tensor(2.5000)

In [7]: t.shape
Out[7]: torch.Size([])
```

pab-vmware commented:
JFYI, I will be using my  other account @Abhishek-P to continue with this PR.

pab-vmware commented:
> JFYI, I will be using my other account @Abhishek-P to continue with this PR.

I will be deleting the @pab-vmware  account. 
If the PR stays open with ghost well and good, I will continue adding commits with @Abhishek-P. 
Else I will raise a fresh PR with mention to this
🤞🏽

dirkgr commented:
No problem, either way is fine. Let me know when you think it's ready for another review!

Abhishek-P commented:
- [x]  SQuAD 1.1 - https://huggingface.co/datasets/squad#data-instances - ref: https://github.com/huggingface/datasets/discussions/2785 for help. -  Deal with answers as a **Dictionary of Lists** - https://gist.github.com/Abhishek-P/ff99afe3dca4302ac1da301a7f3270cf
- [ ] SQuAD 2.0 - 
- [ x] CoNLL 2003 - https://huggingface.co/datasets/conll2003 - https://gist.github.com/Abhishek-P/df28429acbd7408ed74652f18eb64b4e
- [x] sst - https://gist.github.com/Abhishek-P/65e58f44f221f2e941163e6aa483ced8
- [ ] openwebtext
- [ ] MOCHA
- [ ] https://huggingface.co/datasets/commonsense_qa
- [x] https://huggingface.co/datasets/piqa - https://gist.github.com/Abhishek-P/eb2a525ce958ffd0cd0412657648df5c
- [ ] https://huggingface.co/datasets/swag
- [ ] https://huggingface.co/datasets/snli
- [ ] https://huggingface.co/datasets/multi_nli
- [x] https://huggingface.co/datasets/glue - https://gist.github.com/Abhishek-P/65e58f44f221f2e941163e6aa483ced8
- [ ] https://huggingface.co/datasets/super_glue
- [ ] https://huggingface.co/datasets/drop

wolhandlerdeb commented:
Hi, 
I would be glad to know if you add this code to the allennlp library. It would be very helpful.
Thanks

Abhishek-P commented:
> Hi, I would be glad to know if you add this code to the allennlp library. It would be very helpful. Thanks

Hi @wolhandlerdeb, 
 I am a little strapped for bandwidth and hence this has been at the bottom of my personal queue. 
It is pending some work, before it is up to standards for check-in as described in https://github.com/allenai/allennlp/pull/5095#issuecomment-820025555

Please feel free to let clone the changes and continue the work. 
I plan to get back to it in some time, although I am not exactly sure when as of yet.
